### ISSUE TITLE
[bnd] Add postgenerate function

### ISSUE THREAD

**Issue by kriegfrj** [bndtools/bnd#4802]
I had an idea that I wanted to float here for discussion.

I had a use case where I wanted to generate sources, however I needed access to some of the annotations in order to generate the source. This means that the regular generate functionality added by @pkriens was not immediately useful.

I tried using a standard Java annotation processor, but I had issues with that approach too (unfortunately I don't recall the exact reasons; this was a while ago, however I think it may have been due to a lack of support for the annotation processor lifecycle phase in the bnd workspace model that I am using).

So I had an idea to have a "postgenerate" plugin. This would be implemented as an Analyzer plugin and would contain boilerplate code to set up the Java compiler API and invoke it on the generated source. The generated class file(s) (and optionally the generated sources too) would be directly added to the output jar. Specific implementations would subclass this postgenerate plugin and provide the custom code needed to generate the sources, and the base class' boilerplate code would take care of compiling it and ensuring that the source and class files are added to the correct place in the jar.

I think that the regular annotation processor approach could also be made work for my use case. But I think there are some distinct advantages to the approach suggested above: you have the full power and information that Bnd's classfile analysis provides available to you to generate your source.

I don't see this as a replacement for the current generate support - there are times when your hand-coded source depends on the generated source, so you need it to be generated prior to the regular Java build starting. However, in my use case I have the opposite constraint as the generated classes depend on the hand-coded source.

Thoughts?

**Comment by pkriens** [comment]
I think it is interesting. I guess generating the code is quite easy with the compiler API. 

In your case, you could just parse the Java source code with the annotations? A bit inefficient, I agree, but I think it could work?

**Comment by bjhargrave** [comment]
> So I had an idea to have a "postgenerate" plugin. This would be implemented as an Analyzer plugin and would contain boilerplate code to set up the Java compiler API and invoke it on the generated source.

How will you know what Java compiler to use and what options? Especially in the command line (gradle) build.

The current generate support works because Bnd hooks in before compilation (before compilation participant in Eclipse and a gradle task which compileJava depends upon). To add a second compilation step is much more complicated. How will you get Eclipse to compile the generated source code without retriggering the generation of the source code? In gradle, you would need to add a new SourceSet for the generated code and the new SourceSet will need to depend on the main SourceSet (like I had to do in the gradle-plugins project for the Kotlin DSL code).



> In your case, you could just parse the Java source code with the annotations? A bit inefficient, I agree, but I think it could work?

This may be a better solution since your generation is before the main compilation step. Adding a second compilation step is very awkward and cumbersome.

**Comment by kriegfrj** [comment]
> > So I had an idea to have a "postgenerate" plugin. This would be implemented as an Analyzer plugin and would contain boilerplate code to set up the Java compiler API and invoke it on the generated source.
> 
> How will you know what Java compiler to use and what options? Especially in the command line (gradle) build.
> 
> The current generate support works because Bnd hooks in before compilation (before compilation participant in Eclipse and a gradle task which compileJava depends upon). To add a second compilation step is much more complicated. How will you get Eclipse to compile the generated source code without retriggering the generation of the source code? In gradle, you would need to add a new SourceSet for the generated code and the new SourceSet will need to depend on the main SourceSet (like I had to do in the gradle-plugins project for the Kotlin DSL code).
>

I don't think it would be that complicated. The plan was that the Analyzer plugin would generate the source code *and* compile it using `javax.tools` (using the containing project's classpath). So the compilation would be an implementation detail of the Bnd plugin and not an additional compilation stage that the build tool (Eclipse/Gradle/Maven) would need to be aware of at a macro level. In this sense, it's actually cleaner than the bnd generate option, which (correct me if I'm wrong @pkriens) requires Bnd to run before *and* after the build tool's compile phase (before to generate the source, after to do everything that Bnd normally does to make the bundle).

In many ways, this is less complicated than the pre-generate phase which must integrate with the build tool's own compile step. The most difficult part would be reporting compile errors in the generated source in a sensible way, as there would be no source to hang the markers on.

(Btw, for posterity: I just had another look at the annotation processor approach, and I think that the main thing stopping me from using an annotation processor is the difficulties around maintaining build integrity between Eclipse and Gradle in the Bnd workspace model.)

> > In your case, you could just parse the Java source code with the annotations? A bit inefficient, I agree, but I think it could work?
> 
> This may be a better solution since your generation is before the main compilation step. Adding a second compilation step is very awkward and cumbersome.

I might be able get something like this to work. But given all the work that needs to go into finding all the files and parsing them, it might not be much easier than the more generic and more powerful alternative that I had proposed.

Actually, I'm just looking into the annotation processor interface again and I realised the reason I started contemplating this option - I think I *could* do what I want to do with an annotation processor, but it would take a bit of work to integrate it with the Bnd workspace model. At first glance it looked like it might be easier to do it as an analyzer plugin, plus you have the added benefit of having the power of Bnd to analyse the classes.
